lib/vex/validators/type.ex

```diff
@@ -0,0 +1,138 @@
+    unless_skipping(value, options) do
+      acceptable_types = Keyword.get(options, :is, [])
+      if do_validate(value, acceptable_types) do
+        :ok
+      else
+        message = "must be of type #{acceptable_type_str(acceptable_types)}"
+        {:error, message(options, message, value: value)}
+      end
+    end
+  end
+
+  # Allow any type, useful for composed types
+  defp do_validate(_value, :any), do: true
+
+  # Simple types
+  defp do_validate(value, :atom)      when is_atom(value),      do: true
+  defp do_validate(value, :number)    when is_number(value),    do: true
+  defp do_validate(value, :integer)   when is_integer(value),   do: true
+  defp do_validate(value, :float)     when is_float(value),     do: true
+  defp do_validate(value, :binary)    when is_binary(value),    do: true
+  defp do_validate(value, :bitstring) when is_bitstring(value), do: true
+  defp do_validate(value, :tuple)     when is_tuple(value),     do: true
+  defp do_validate(value, :list)      when is_list(value),      do: true
+  defp do_validate(value, :map)       when is_map(value),       do: true
+  defp do_validate(value, :function)  when is_function(value),  do: true
+  defp do_validate(value, :reference) when is_reference(value), do: true
+  defp do_validate(value, :port)      when is_port(value),      do: true
+  defp do_validate(value, :pid)       when is_pid(value),       do: true
+  defp do_validate(%{__struct__: module}, module),              do: true
+
+  # Complex types
+  defp do_validate(value, :string) when is_binary(value) do
+    String.valid?(value)
+  end
+
+  defp do_validate(value, :regex) do
+    Regex.regex?(value)
+  end
+
+  defp do_validate(value, function: arity) when is_function(value, arity), do: true
+
+  defp do_validate(list, list: type) when is_list(list) do
+    Enum.all?(list, &(do_validate(&1, type)))
+  end
+  defp do_validate(value, map: {key_type, value_type}) when is_map(value) do
+    Enum.all? value, fn {k, v} ->
+      do_validate(k, key_type) && do_validate(v, value_type)
+    end
+  end
+  defp do_validate(tuple, tuple: types)
+      when is_tuple(tuple) and is_tuple(types) and tuple_size(tuple) == tuple_size(types) do
+    Enum.all? Enum.zip(Tuple.to_list(tuple), Tuple.to_list(types)), fn {value, type} ->
+      do_validate(value, type)
+    end
+  end
+
+  # Accept multiple types
+  defp do_validate(value, acceptable_types) when is_list(acceptable_types) do
+    Enum.any?(acceptable_types, &(do_validate(value, &1)))
+  end
+
+  # Fail if nothing above matched
+  defp do_validate(_value, _type), do: false
+
+
+  defp acceptable_type_str([acceptable_type]), do: inspect(acceptable_type)
+  defp acceptable_type_str(acceptable_types) when is_list(acceptable_types) do
+    last_type = acceptable_types |> List.last |> inspect
+    but_last =
+      acceptable_types
+      |> Enum.take(Enum.count(acceptable_types) - 1)
+      |> Enum.map(&inspect/1)
+      |> Enum.join(", ")
+    "#{but_last} or #{last_type}"
```

Comment:
Can you elaborate a bit about what this function is doing? I get what the result is but the method you need to achieve it is confusing to me.